pkg/config/initoptions.go

```diff
@@ -0,0 +1,179 @@
+// Copyright 2020 The Kubernetes Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package config
+
+import (
+	"fmt"
+	"math/rand"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"time"
+
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"sigs.k8s.io/kustomize/kyaml/kio"
+)
+
+const manifestFilename = "grouping-object-template.yaml"
+
+const configMapTemplate = `# NOTE: auto-generated
+# Date: <DATETIME>
+#
+# Contains the "grouping object" template ConfigMap.
+# When this object is applied, it is handled specially,
+# storing the metadata of all the other objects applied.
+# This object and its stored inventory is subsequently
+# used to calculate the set of objects to automatically
+# delete (prune), when an object is omitted from further
+# applies.
+#
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  namespace: <NAMESPACE>
+  name: grouping-object
+  labels:
+    # This label defines the ConfigMap as a grouping object template
+    cli-utils.sigs.k8s.io/inventory-id: <GROUPNAME>
+`
+
+// InitOptions contains the fields necessary to generate a
+// grouping object template ConfigMap.
+type InitOptions struct {
+	ioStreams genericclioptions.IOStreams
+	// Package directory argument
+	Dir string
+	// Namespace for grouping object
+	Namespace string
+	// Grouping object label value
+	GroupName string
+	// Random seed
+	Seed int64
+}
+
+func NewInitOptions(ioStreams genericclioptions.IOStreams) *InitOptions {
+	return &InitOptions{
+		ioStreams: ioStreams,
+	}
+}
+
+// Complete fills in the InitOptions fields.
+func (i *InitOptions) Complete(args []string) error {
+	i.Seed = time.Now().UnixNano()
+	if len(args) != 1 {
+		return fmt.Errorf("need one 'directory' arg; have %d", len(args))
+	}
+	i.Dir = args[0]
+	if !isDirectory(i.Dir) {
+		return fmt.Errorf("invalid directory argument: %s", i.Dir)
+	}
+	if len(i.Namespace) == 0 {
+		namespace, err := packageNamespace(i.Dir)
+		if err != nil {
+			return err
+		}
+		i.Namespace = namespace
+	}
+	if len(i.GroupName) == 0 {
+		i.GroupName = i.defaultGroupName()
+	}
+	if !validateGroupName(i.GroupName) {
+		return fmt.Errorf("invalid group name: %s", i.GroupName)
+	}
+	return nil
+}
+
+// isDirectory returns true if the passed path is a directory;
+// false otherwise.
+func isDirectory(path string) bool {
+	if d, err := os.Stat(path); err == nil {
+		if d.IsDir() {
+			return true
+		}
+	}
+	return false
+}
+
+// packageNamespace returns the namespace of the package
+// config files.
+func packageNamespace(packageDir string) (string, error) {
+	r := kio.LocalPackageReader{PackagePath: packageDir}
+	nodes, err := r.Read()
+	if err != nil {
+		return "", err
+	}
+	namespace := "default"
+	for _, node := range nodes {
+		rm, err := node.GetMeta()
+		if err != nil {
+			continue
+		}
+		if len(rm.ObjectMeta.Namespace) > 0 {
+			namespace = rm.ObjectMeta.Namespace
+		}
+		break
```

Comment:
So we just take the namespace from the first resource we find? Isn't that a bit fragile?

Reply:
I don't think so. Either we allow multiple namespaces or we don't. If we don't then it will be fine, and the namespace validation will catch any problems. If we do, then it also works, since we just want the grouping object in one of the namespaces.

Reply:
But do we ever check whether there are multiple namespaces in the package?
Just remembered this now, but I would assume that the namespace is a field that will usually be defined with a setter. Fetching it from the manifests could lead to surprising results if someone uses the setter after running init.

Reply:
Good point. It can also be set with Kustomize build.